### ISSUE TITLE
Bad display with agglite driver for a given shape

### ISSUE THREAD

**Issue by JD31** [gwaldron/osgearth#798]
Here is the layer:

```
<image name="world_boundaries" driver="agglite" visible="true">

    <features name="world" driver="ogr">
        <connection>PGB:dbname='pgdump_1607' host='localhost' port='5432' user='postgres'</connection>
        <layer>world2</layer>
        <ogr_driver>PostgreSQL</ogr_driver>
        <build_spatial_index>true</build_spatial_index>
    </features>


    <styles>
        <style type="text/css">
            default {
            altitude-clamping:  terrain;
            altitude-technique: drape;
            fill:               #373737;
            stroke:             #000000;
            stroke-width:       2;
            }
        </style>
    </styles>
</image> 
```

Some countries are displayed black...
(in the above file the db connection must be replaced with the given shape file)
Issue is with 2.8 RC1, GDal 2.1.0, Osg 3.4, MSVC2015 64b

[countries.zip](https://github.com/gwaldron/osgearth/files/464327/countries.zip)

**Comment by jasonbeverage** [comment]
I can reproduce this, looking into it now.  Is this something has broken recently or has this always happened?

**Comment by JD31** [comment]
It is difficult to say because i used to drape this model.
However i have already tried agglite and it was ok. I think it was at the
beginning of this year.

**Comment by jasonbeverage** [comment]
The issue appears to be that the agglite driver is having issues with the MultiPolygons in the file.  The polygons that just turn black are b/c there are polygons instead of rings in the line geometry for some reason.

**Comment by JD31** [comment]
Mmmmmm...
But why the black polygons are changing when changing camera altitude ?
Le 9 sept. 2016 7:22 PM, "Jason Beverage" <notifications@github.com> a
écrit :

**Comment by jasonbeverage** [comment]
Unfortunately there are a host of issues here that we aren't going to have time to fix before osgEarth 2.8, but we will keep this on our radar.

The first and most obvious issue is that the cloneAs class here won't work for MultiPolygons, which your dataset has tons of.
https://github.com/gwaldron/osgearth/blob/master/src/osgEarthDrivers/agglite/AGGLiteRasterizerTileSource.cpp#L195

Even after getting around that issue by just creating lines for all of the components of the multipolygons, there are still a host of geos related issues where it's not able to crop and or buffer some of your geometries, resulting in crashes and incorrect rendering.

I've pushed a branch called [fixagglite](https://github.com/gwaldron/osgearth/tree/fixagglite) that experiments with rendering the lines without doing any buffering, but it's so slow with your dataset that it's borderline unusable.  

We'll have to revisit this later, but thanks again for providing the test case, it'll be very useful for coming up with a proper solution.

**Comment by JD31** [comment]
Hi. Can we expect a correction for osgEarth 2.9?

**Comment by jasonbeverage** [comment]
Hi JD, this should be fixed now.  Thanks for the push :)

I changed the aaglite driver to use the GeometryIterator for collecting lines since we can't do a cloneAs with a multipolygon.  This is the same fix as I did before when I was testing, but now there doesn't appear to be any buffer errors or crashes as last time I looked at this.  Something may have been fixed along the way in the buffering or geos code, but it's looking good now.
